Rx/v2/src/rxcpp/operators/rx-debounce.hpp

```diff
@@ -7,209 +7,299 @@
 
 #include "../rx-includes.hpp"
 
+/*! \file rx-debounce.hpp
+
+    \brief  Return an observable that emits an item if a particular timespan has passed without emitting another item from the source observable.
+
+    \tparam Duration      the type of time interval
+    \tparam Coordination  the type of the scheduler
+
+    \param period        the period of time to suppress any emitted items
+    \param coordination  the scheduler to manage timeout for each event
+
+    \return  Observable that emits an item if a particular timespan has passed without emitting another item from the source observable.
+
+    \sample
+    \snippet debounce.cpp debounce sample
+    \snippet output.txt debounce sample
+*/
+
 namespace rxcpp {
 
 namespace operators {
 
 namespace detail {
 
-template<class T, class Duration, class Coordination>
-struct debounce
-{
-    typedef rxu::decay_t<T> source_value_type;
+template<class... AN>
+struct debounce_invalid_arguments {};
+
+template<class... AN>
+struct debounce_invalid : public rxo::operator_base<debounce_invalid_arguments<AN...>> {
+    using type = observable<debounce_invalid_arguments<AN...>, debounce_invalid<AN...>>;
+};
+template<class... AN>
+using debounce_invalid_t = typename debounce_invalid<AN...>::type;    
+    
+    
+template <class Coordination, class Duration, class Observable>
+struct debounce_traits {
+    typedef rxu::decay_t<Duration> duration_type;
     typedef rxu::decay_t<Coordination> coordination_type;
+    typedef rxu::decay_t<Observable> source_type;
+    typedef typename source_type::value_type source_value_type;
+    typedef source_value_type value_type;
+};
+
+template<class Coordination, class Duration, class Observable>
+struct debounce : public operator_base<rxu::value_type_t<debounce_traits<Coordination, Duration, Observable>>>
+{
```

Comment:
I am trying out a pattern for the lifters in #260 
Hopefully VS will compile :)